### DISCUSSION TITLE
ToDictionary TKey nullable check

### DISCUSSION THREAD

**Discussion by IgorMenshikov** [dotnet/csharplang#4608] [Q&A] [answered]
I have a linq query
```C#
var periodHash = context.Query<Entity>()
                .Where(i => i.Name != null)
                .ToDictionary(i => i.Name, i => i);
```
So, I can be sure key cannot be null.

But I get a warning:
```
Warning	CS8714	The type 'string?' cannot be used as type parameter 'TKey' in the generic type or method 'Enumerable.ToDictionary<TSource, TKey, TElement>(IEnumerable<TSource>, Func<TSource, TKey>, Func<TSource, TElement>)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
```

Is there a way to avoid the warning?

**Comment by dstarkowski** [comment, marked as answer]
Either declare `Name` in `Entity` class as non nullable, or use `!` to tell compiler `i.Name` is not null.

    var periodHash = context.Query<Entity>()
        .Where(i => i.Name != null)
        .ToDictionary(i => i.Name!, i => i);
Also see #3951

**Reply by IgorMenshikov** [reply]
1. Name can be nullable.  That is why I filter null values out.
2. Null-forgiving operator helps to hide the warning. Thanks

**Reply by FaustVX** [reply]
@IgorMenshikov Just FYI, the `Where` method don't change the type.
So, `Name` type don't change from `string?` to `string`.
And so, in `.ToDictionary(i => i.Name, i => i)`, `TKey` is still `string?`.

**Reply by IgorMenshikov** [reply]
This is more about "null warning logic". If you write something like:
```C#
string? something = GetSomething();
if (something != null) {
    something.ToString(); // no warning here because of data flow.
}
```

Same logic would be good for Where. It is like "if" for collection.

But, we have a workaround using Null-forgiving operator.

**Reply by Joe4evr** [reply]
@IgorMenshikov That's literally what #3951 is about.
